### PR TITLE
Mir audit: Batchfetcher and mempool modules update to DSL codegen.

### PR DESCRIPTION
**Pull request by ranchalp**

### CODE REVIEW

**Comment by ranchalp** [comment]
Builds upon #336

**Comment by ranchalp** [comment]
Added a fix for #353 , the fix may interfere with `mpdsl.UponRequestTransactions` if that call is for a number of txIDs that are not in memory anymore, but I could not find a single event being created for this purpose.

**Comment by ranchalp** [comment]
Also, found #352 along the way. We should discuss how much of an issue we think that could be (specially with the latest's availability module #336 .

**Comment by ranchalp** [comment]
Forced push to account for possibility of multiple client's requests with the same txID when fixing #353.

**Comment by ranchalp** [comment]
Added support for map types to codegen, although it is not being used at the moment. The mir.type field must define both key and val types in the following way `keytype|valtype`

**Comment by matejpavlovic** [comment]
I'm putting this on hold until #336 is polished and merged, let's have a look at this one afterwards.

File: pkg/batchfetcher/batchfetcher.go
Comment (matejpavlovic) [inline pkg/batchfetcher/batchfetcher.go:113]: I'd actually prefer using the checkpoint abstraction. It might be marginally les performant, but I find it cleaner.

File: pkg/batchfetcher/batchfetcher.go
Comment (matejpavlovic) [inline pkg/batchfetcher/batchfetcher.go:112]: If we keep using the checkpoint abstraction, we don't need to rely on the internal format of the data structure (in case it ever changes). And the code will also be easier to look at.

File: codegen/model/types/map.go
Comment (matejpavlovic) [inline codegen/model/types/map.go:11]: Forgot copy-pasted comment?

**Review by matejpavlovic — COMMENTED** [review]
Great job! Nicely structured PR.
Even better, adding the map support to the codegen could have been done in a separate PR to make it easy to review separately. In fact, even the updates to the batch fetcher and mempool could go to different PRs next time. Still, very nice job!!

Let's try to get @xosmig 's review on the map support in the code generation, I'm not too familiar with the code and it will probably take him 5 minutes.

Andrei, would you have time to look at the codegen part? It's just 2 files and probably it won't even take you 5 minutes. Thanks a lot!

File: codegen/model/types/parser.go
Comment (xosmig) [inline codegen/model/types/parser.go:245]: Probably can be removed now :)

**Review by xosmig — APPROVED** [review]
I added a few notes regarding the type annotation syntax, but other than that, the codegen part looks great, thank you! :)

File: codegen/model/types/parser.go
Comment (xosmig) [inline codegen/model/types/parser.go:284]: Also, I would be curious to know your opinion on potential alternative syntaxes for the annotation.

A few natural examples (in no specific order):
1. using -> as the separator and trimming the whitespaces around it
2. using map<keyType, valueType> to follow the syntax of protobufs
3. using map[keyType]valueType to follow the syntax of Go
4. using two separate annotations (mir.key_type) and (mir.value_type) and throwing an error if (mir.type) is used for a map type

One potential concern about all options except the last one (including the current version) is the length of the annotation, since, as of now, the type annotation has to include the full package path for each type.

Comment (xosmig) [reply codegen/model/types/parser.go:284]: Option (4) could look somewhat like this:

```protobuf
message MapMessage {
    map<string, int> mapValue [
        (mir.key_type) = "github.com/awesome-company/awesome-project/package.KeyType",
        (mir.value_type) = "github.com/awesome-company/awesome-project/package.ValueType",
    ];
}
```

The current version may be a bit hard to parse visually:

```protobuf
message MapMessage {
    map<string, int> mapValue [(mir.type) = "github.com/awesome-company/awesome-project/package.KeyType|github.com/awesome-company/awesome-project/package.ValueType"];
}
```

It could probably be helped to a certain extent by trimming the whitespaces around the separator.

**Comment by matejpavlovic** [comment]
Thanks a lot @xosmig for the comments, it's very helpful!

File: codegen/model/types/parser.go
Comment (matejpavlovic) [reply codegen/model/types/parser.go:284]: Intuitively, option (4) seems the nicest to me. On top of what you already said, it seems to avoid the necessity of string parsing magic. @ranchalp what do you think?

File: pkg/batchfetcher/batchfetcher.go
Comment (ranchalp) [reply pkg/batchfetcher/batchfetcher.go]: This is something I have been thinking about it myself. I will add the check as it is a (small) optimization, but I think we should open a TODO/issue for a more standard way to check for empty certificates (in the view change it is a `[]byte{}`, in the availability module it is `[]byte{0}` and here it seems to be a nil. This is not immediate though as it might even make sense to have different empty certs depending on which module should check.

File: pkg/batchfetcher/batchfetcher.go
Comment (ranchalp) [reply pkg/batchfetcher/batchfetcher.go:112]: ditto, need to come back to this to tag the type on protobufs and thus removing `t.EpochNr` once we are back to using the codegen code.

File: pkg/batchfetcher/batchfetcher.go
Comment (ranchalp) [reply pkg/batchfetcher/batchfetcher.go:113]: Done as requested, but may come back to this once I pass through the checkpoint module. Bear in mind that `checkpoint.` refers to a manually written package, whereas `checkpointpbtypes.` is automatically generated. I believe minimizing the amount of manually written code and maximizing tags in protobufs is the way to go. If types change later at least we know that they must be changed just in protobufs.

File: pkg/batchfetcher/batchfetcher.go
Comment (ranchalp) [reply pkg/batchfetcher/batchfetcher.go:115]: ditto.

File: pkg/batchfetcher/events/events.go
Comment (ranchalp) [reply pkg/batchfetcher/events/events.go:11]: yes, done, "treat".

File: pkg/mempool/simplemempool/internal/parts/formbatches/formbatches.go
Comment (ranchalp) [reply pkg/mempool/simplemempool/internal/parts/formbatches/formbatches.go:29]: Yes that is the reason. I am aware that with the current implementation the app logic is responsible for filtering collisions and just keeping one copy, but here the purpose is to prevent the state from blowing app by keeping in memory all transactions (i.e. removing after decision). The fix is clear. However, if I remove a transaction after the first decision and that transaction is proposed again (by a malicious proposer), then, after deciding on this proposal, line `txIDs = append(txIDs, txID)` will throw a runtime error, because `tx := state.TxByID[string(txID)]` will return nil (and `append(slice, nil)` throws a runtime error.

File: codegen/model/types/map.go
Comment (ranchalp) [reply codegen/model/types/map.go:11]: Correct, thanks, fixed.

File: protos/checkpointpb/checkpointpb.proto
Comment (ranchalp) [reply protos/checkpointpb/checkpointpb.proto:36]: It is, but I clarified further, see above explanation.

File: codegen/model/types/parser.go
Comment (ranchalp) [reply codegen/model/types/parser.go:241]: Done (as indicated below)

File: codegen/model/types/parser.go
Comment (ranchalp) [reply codegen/model/types/parser.go:284]: Done

File: codegen/model/types/parser.go
Comment (ranchalp) [reply codegen/model/types/parser.go:245]: Done, the new code also supports defining only the key (by either not writing '|' or writing it as the last character when tagging).

Comment (ranchalp) [reply codegen/model/types/parser.go:245]: Done.

File: codegen/model/types/parser.go
Comment (ranchalp) [reply codegen/model/types/parser.go:284]: Option 4 sounds good. Just did it, and took the extra effort to also support defining only the key or only the value types (which is related to other comments of this PR and will enable for removing even more boilerplate code).

File: pkg/batchfetcher/batchfetcher.go
Comment (ranchalp) [reply pkg/batchfetcher/batchfetcher.go:112]: Update: after addressing @xosmig 's comments I decided to also add support for map types defining only the key or the value type. So in a new upcoming PR we can clean up this code further. Not doing it on this PR for the purpose of not overcomplicating it at this stage.

File: codegen/model/types/parser.go
Comment (ranchalp) [reply codegen/model/types/parser.go:284]: I also left as TODO in issue #356 the last bit that I think we need to completely convert all proto messages with the codegen. Will be doing that in an upcoming PR. Thanks for your comments! :)

**Comment by ranchalp** [comment]
Thanks @xosmig and @matejpavlovic for the comments! :)

File: pkg/mempool/simplemempool/internal/parts/formbatches/formbatches.go
Comment (matejpavlovic) [reply pkg/mempool/simplemempool/internal/parts/formbatches/formbatches.go:29]: I think there are 2 concepts being mixed here: (1) proposing transactions and (2) adding transactions to the mempool.

Unless I missed something (possible), the mempool is a purely local data structure that serves as a buffer for transactions waiting to be proposed by _this node_, regardless of other malicious nodes are proposing. Or does the mempool get involved in any way during other nodes' proposals? It is possible, though, that some client (not even necessarily a malicious one) can send the same transaction to the same node more than once.

But this should be addressed differently in my opinion. It would be both more natural and way simpler to just keep at most one copy of a transaction in the mempool.
1. Remove the counter
2. When adding transactions, ignore those that are already inside according to the index (note that this is easy, as everything is executed sequentially)

With this approach, is there still a problem that the counter would solve?

File: codegen/model/types/map.go
Comment (matejpavlovic) [reply codegen/model/types/map.go]: Is it? 😄

File: pkg/batchfetcher/batchfetcher.go
Comment (ranchalp) [reply pkg/batchfetcher/batchfetcher.go:130]: You are right, this is not necessary anymore, removing...

File: pkg/mempool/simplemempool/internal/parts/formbatches/formbatches.go
Comment (ranchalp) [reply pkg/mempool/simplemempool/internal/parts/formbatches/formbatches.go:29]: Actually yes you are absolutely right, I thought the mempool was storing transactions received from batches by the availability module, which in this case is of course the responsibility of the fakebatchdb (and obviously the fakebatchdb is already known to blow up the state by keeping the database in plain maps in memory). I just updated the code to remove the counter on the txID (and assume as elsewhere that the app logic prevents collisions).

Comment (matejpavlovic) [reply pkg/mempool/simplemempool/internal/parts/formbatches/formbatches.go]: What does need to be fixed, though, is checking for the presence of the transaction when adding it, and only add it if it is not yet in the `TxByID` index (otherwise ignore it).

File: codegen/model/types/parser.go
Comment (xosmig) [reply codegen/model/types/parser.go]: Sorry, I just realized that there is actually a perfectly valid use-case for using all three of `(mir.value_type)`, `(mir.key_type)`, and `(mir.type)`. Namely, when there is a wrapped map type.

Something like:
```go
// Membership represents the list of participants with their respective weights.
type Membership map[NodeID]int
```

Comment (xosmig) [reply codegen/model/types/parser.go]: ah, but maybe it's already supported by the current code :)

File: pkg/mempool/simplemempool/internal/parts/formbatches/formbatches.go
Comment (ranchalp) [reply pkg/mempool/simplemempool/internal/parts/formbatches/formbatches.go]: done

File: codegen/model/types/parser.go
Comment (xosmig) [reply codegen/model/types/parser.go]: No, I guess it's not, but it should be the matter of removing the extra check

File: codegen/model/types/parser.go
Comment (xosmig) [inline codegen/model/types/parser.go:278]:
```suggestion
```

File: codegen/model/types/parser.go
Comment (xosmig) [reply codegen/model/types/parser.go]: Hm, there is an issue though: it won't be consistent with how `(mir.type)` operates on slices

Comment (xosmig) [reply codegen/model/types/parser.go]: maybe, in the long term (perhaps, not in this PR), it would be better to migrate slices to using a pair of annotations:
1. `(mir.value_type)` for the underlying type
2. `(mir.type)` for wrapping the slice itself

Comment (matejpavlovic) [reply codegen/model/types/parser.go]: Could that be somehow related to a solution to #356 ?

But I agree that this should definitely be addressed in a different PR.

**Pull request merged**
